main/acle.md

```diff
@@ -4574,6 +4579,406 @@ two pointers, ignoring the tags.
+
+The table below summarizes the possible ZA states on entry to an
+inline asm A in a function F. It also specifies what the state
+is required to be when A finishes; it is the asm's responsibility
+to ensure this. Since PSTATE.ZA controls both ZA and ZT0, the rules
+depend on ZT0 as well as ZA.
+
+| **ZA state before A** | **ZA state after A** | **Possible if…**                       |
+| --------------------- | -------------------- | -------------------------------------- |
+| off                   | off                  | F's uses and A's clobbers are disjoint |
+| dormant               | dormant              | " " "                                  |
+| dormant               | off                  | " " ", and A clobbers `"za"`           |
+| active                | active               | F uses `"za"` and/or `"zt0"`           |
+
+The [`__ARM_STATE` macros](#state-strings) indicate whether a compiler
+is guaranteed to support a particular clobber string. For example,
+the following code is a safe way to indicate that an asm might commit
+a lazy ZA save:
+
+``` c
+  // Function foo doesn't use ZA or ZT0.
+  void foo() {
+    ...
+    asm volatile("bl something"
+                 ::: ...call-clobbered registers...
+#ifdef __ARM_STATE_ZA
+                     , "za"
+#endif
+                );
+    ...
+  }
+```
+
+## Implementation notes
+
+<span id="za-implementation-notes"></span>
+
+A piece of state S can be said to be “live” at a point P during the
+execution of a function F if:
+
+*   F [uses](#uses-state) S; and
+
+*   a random, exogenous change to S at P could change the behavior
+    of the program.
+
+If S is live before a call from F to a function F2 that does not share S
+with its caller, the compiler must arrange for S to be preserved around
+the call to F2. One way to do this is to save S before the call and
+restore it afterwards. However, if S is ZA and F2 is a private-ZA
+function, the compiler can instead make use of the lazy-save scheme
+described in [[AAPCS64]](#AAPCS64).
+
+For example, the code below shows a function that uses ZA and ZT0
+internally. The comments describe when ZA should be saved and restored:
+
+``` c
+  void f_shares_zt0() __arm_inout("zt0");
+  void f_shares_za() __arm_inout("za");
+  void f_private();
+
+  void setup_za() __arm_out("za");
+  void use_za() __arm_in("za");
+
+  __arm_new("za", "zt0") void test() {
+    f_private();     // ZA is not live, no save necessary
+
+    setup_za();      // ZA is live after this call
+    f_shares_zt0();  // The compiler should save and restore ZA
+                     // around the call ("caller-save")
+    f_shares_za();   // ZA is live before and after this call
+    f_private();     // The compiler should preserve ZA across the call
+                     // It can use the lazy-save mechanism
+    use_za();        // ZA is no longer live after this call
+
+    f_private();     // ZA is not live, no save necessary
```

Comment:
Could you also describe the liveness of ZT0 (and the need to save/restore it) as part of the example?

Reply:
Good idea.  In the end it seemed simpler to use a separate example, since it became a bit unwieldly otherwise.